agent.go

```diff
@@ -50,80 +55,76 @@ func InstallNewVersion(ctx context.Context, logger logging.Logger) (bool, error)
+	expectedBinPath := filepath.Join(utils.ViamDirs.Bin, SubsystemName)
+	arch := utils.GoArchToOSArch(runtime.GOARCH)
+	if arch == "" {
+		return fmt.Errorf("could not determine platform arch mapping for GOARCH %s", runtime.GOARCH)
+	}
+	expectedCachePath := filepath.Join(utils.ViamDirs.Cache, strings.Join([]string{SubsystemName, "v" + utils.Version, arch}, "-"))
 	curPath, err := os.Executable()
 	if err != nil {
 		return errw.Wrap(err, "getting path to self")
 	}
 
-	isSelf, err := utils.CheckIfSame(curPath, expectedPath)
+	isExpected, err := utils.CheckIfSame(expectedCachePath, expectedBinPath)
 	if err != nil {
-		return errw.Wrap(err, "checking if installed viam-agent is myself")
+		return errw.Wrap(err, "checking if installed viam-agent is in expected state")
 	}
 
-	if !isSelf {
-		logger.Infof("adding a symlink to %s at %s", curPath, expectedPath)
-		if err := os.Remove(expectedPath); err != nil && !errw.Is(err, fs.ErrNotExist) {
-			return errw.Wrapf(err, "removing symlink/file at %s", expectedPath)
+	if !isExpected {
+		logger.Infof("installing to %s and adding a symlink at %s", expectedCachePath, expectedBinPath)
+		err := utils.AtomicCopy(expectedCachePath, curPath)
+		if err != nil {
+			return errw.Wrap(err, "installing self into cache directory")
 		}
-		if err := os.Symlink(curPath, expectedPath); err != nil {
-			return errw.Wrapf(err, "installing symlink at %s", expectedPath)
+		if err := os.Remove(expectedBinPath); err != nil && !errw.Is(err, fs.ErrNotExist) {
+			return errw.Wrapf(err, "removing symlink/file at %s", expectedBinPath)
 		}
-	}
-
-	serviceFilePath, removeOldFile, err := getServiceFilePath(logger)
-	if err != nil {
-		return errw.Wrap(err, "getting service file path")
-	}
-
-	// use this later to avoid re-enabling an existing agent service a user might have disabled
-	_, err = os.Stat(serviceFilePath)
-	newInstall := err != nil
-
-	logger.Infof("writing systemd service file to %s", serviceFilePath)
-
-	newFile, err := utils.WriteFileIfNew(serviceFilePath, serviceFileContents)
-	if err != nil {
-		return errw.Wrapf(err, "writing systemd service file %s", serviceFilePath)
-	}
-
-	if removeOldFile {
-		oldPath := filepath.Join(fallbackFileDir, serviceFileName)
-		logger.Warn("Removing system service file %s in favor of vendor file at %s", oldPath, serviceFilePath)
-		logger.Warn("If you customized this file, please run 'systemctl edit viam-agent' and create overrides there")
-		if err := os.RemoveAll(oldPath); err != nil {
-			logger.Warn(errw.Wrapf(err, "removing old service file %s, please delete manually", oldPath))
+		if err := os.Symlink(expectedCachePath, expectedBinPath); err != nil {
+			return errw.Wrapf(err, "installing symlink at %s", expectedBinPath)
 		}
-	}
 
-	if newFile {
-		cmd = exec.Command("systemctl", "daemon-reload")
-		output, err = cmd.CombinedOutput()
-		if err != nil {
-			return errw.Wrapf(err, "running 'systemctl daemon-reload' output: %s", output)
+		if !VersionCacheExists() {
```

Comment:
This conditional is the only functional change from the last PR. In the case of a self update the previous version of the agent that's performing the update handles updating the cache, so we don't need to touch it.